### PR TITLE
Make secagg insecure validation optional

### PR DESCRIPTION
**Pull request by mvesin**
**PR description**

- implement #804 for making secagg insecure validation optional
- misc: fix for secagg end2end test & `DefaultStrategy()` ; adapt VPN documentation for LOM secagg ; update `Environ` docstrings

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree to the
[Developer Certificate of Origin (DCO)](https://github.com/fedbiomed/fedbiomed/blob/master/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.

**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.

**Guidelines for PR review**

General:

* give a glance to [DoD](https://fedbiomed.org/latest/developer/definition-of-done/)
* check [coding rules and coding style](https://fedbiomed.org/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state_breakpoint`/`load_state_breakpoint` in aggregators, strategies, secagg, etc.)
* if modified a component with versioning (config files, breakpoint, messaging protocol) then update the version following the rules in `common/utils/_versions.py`

### CODE REVIEW

File: fedbiomed/common/environ.py [hunk 1]
```diff
@@ -183,3 +168,10 @@ def set_environment(self):
             "MPSPDZ_CERTIFICATE_PEM",
             os.path.join(self._values["CONFIG_DIR"], public_key)
         )
+
+        # make `secagg_insecure_validation` optional in config file (and avoid changing config file version)
+        # by using a `fallback` value
+        secagg_insecure_validation = self.config.get('security', 'secagg_insecure_validation', fallback='true')
```
Comment (srcansiz) [inline fedbiomed/common/environ.py:174]: Please find the suggestion for reducing the size of the line and a shorter comment.  

```suggestion
        # Optional secagg_insecure_validation optional in config file
        secagg_insecure_validation = self.config.get(
            'security', 'secagg_insecure_validation', fallback='true')
```

File: fedbiomed/node/round.py [hunk 2]
```diff
@@ -350,11 +358,14 @@ def run_model_training(
...
-                    weight=results['sample_size'],
-                )
+                if self._secure_aggregation.scheme.secagg_random is not None:
+                    results["encryption_factor"] = self._secure_aggregation.scheme.encrypt(
+                        params=[self._secure_aggregation.scheme.secagg_random],
+                        current_round=self._round,
+                        weight=results['sample_size'],
+                    )
+                else:
+                    results["encryption_factor"] = None
```
Comment (srcansiz) [inline fedbiomed/node/round.py:368]: Encryption factor can be `None` by default `results["encryption_factor"]`

```suggestion
                results["encryption_factor"] = None
                if self._secure_aggregation.scheme.secagg_random is not None:
                    results["encryption_factor"] = self._secure_aggregation.scheme.encrypt(
                        params=[self._secure_aggregation.scheme.secagg_random],
                        current_round=self._round,
                        weight=results['sample_size'])                 
```

File: fedbiomed/researcher/secagg/_secure_aggregation.py [hunk 3]
```diff
@@ -306,8 +306,11 @@ def _configure_round(
             experiment_id: The id of the experiment
         """
 
-        # For each round it generates new secagg random float
-        self._secagg_random = round(random.uniform(0, 1), 3)
+        if environ['SECAGG_INSECURE_VALIDATION'] is True:
+            # For each round it generates new secagg random float
+            self._secagg_random = round(random.uniform(0, 1), 3)
+        else:
+            self._secagg_random = None
```
Comment (srcansiz) [inline fedbiomed/researcher/secagg/_secure_aggregation.py:313]:
```suggestion
        self._secagg_random = None
        if environ['SECAGG_INSECURE_VALIDATION'] is True:
            # For each round it generates new secagg random float
            self._secagg_random = round(random.uniform(0, 1), 3)            
```

File: fedbiomed/researcher/secagg/_secure_aggregation.py [hunk 4]
```diff
@@ -331,18 +334,18 @@ def aggregate(
...
-        num_expected_params: int | None = None
+        encryption_factors: Dict[str, Union[List[int], None]],
+        num_expected_params: int | None = None,
     ) -> bool:
         """Validate given inputs"""
 
-
-        if encryption_factors is None:
+        # at this point we know `isinstance(encryption_factors, dict) and len(encryption_factors) == num_nodes`
+        if any([v is None for v in encryption_factors.values()]):
```
Comment (srcansiz) [inline fedbiomed/researcher/secagg/_secure_aggregation.py:343]: if think `all(encryption_factors.values())` does the same think. 

```suggestion
        if not all(encryption_factors.values()):
```

**Review by srcansiz — COMMENTED** [review]
Hi @mvesin,

Thank you for the PR. Generally, the PR is good to go. Please find my suggestions about a few little details.  

I would also suggest to add a basic e2e test to test the error case.

File: fedbiomed/researcher/secagg/_secure_aggregation.py (hunk 4 above)
Comment (mvesin) [reply fedbiomed/researcher/secagg/_secure_aggregation.py:343]: This is not exactly the same test. For example with `encryption_factors.values = [False, [1]]` tests don't return the same value. Testing specifically for `None` value intends to catch only the case of setup mismatch between researcher and node, while doing a `not all()` test will also catch some malicious attack cases, misleading the user to believe this is a setup mismatch (while malicious case are caught in `SecaggCrypter` with explicit message indicating the type of the encryption factors is not correct).

Sio I would prefer to keep the current more explicit test

**Pull request merged**